### PR TITLE
refactor(wininfo): change wininfo from a linked list to an array

### PR DESCRIPTION
**Pull request by bfredl**
"wininfo" is going to be my next victim. The main problem with wininfo is that it is "all or nothing", i e either all state about a buffer in a window is considered valid or none of it is. This needs to be fixed to address some long running grievances.

For now this is just a warmup: refactor it from a linked list to a vector.

### CODE REVIEW

**Review by justinmk — APPROVED** [review]

File: src/klib/kvec.h [hunk 1]
```diff
@@ -135,6 +135,9 @@
          : 0UL)), \
      &(v).items[(i)]))
 
+#define kv_shift(v, i) ((v).size -= 1, (i) < (v).size && memmove(&kv_A(v, (i)), &kv_A(v, (i)+1), \
+                                                                 ((v).size-(i))*sizeof(kv_A(v, i))))
```
Comment (zeertzjq) [inline src/klib/kvec.h:139]: Maybe this should be named `kv_shift1()`, or it should take an argument specifying how far to shift?

**Review by zeertzjq — COMMENTED** [review]
The `FOR_ALL_BUF_WININFO()` defined in `globals.h` is now no longer valid.

File: src/nvim/window.c [hunk 2]
```diff
@@ -5235,30 +5235,22 @@ void win_free(win_T *wp, tabpage_T *tp)
...
-    for (wininfo_T *wip = buf->b_wininfo; wip != NULL; wip = wip->wi_next) {
+    WinInfo *wip_wp = NULL;
+    size_t pos_null = kv_size(buf->b_wininfo);
+    for (size_t i = 0; i < kv_size(buf->b_wininfo); i++) {
+      WinInfo *wip = kv_A(buf->b_wininfo, i);
       if (wip->wi_win == wp) {
-        wininfo_T *wip2;
-
-        // If there already is an entry with "wi_win" set to NULL it
-        // must be removed, it would never be used.
```
Comment (zeertzjq) [inline src/nvim/window.c:5243]: Why is this comment removed?

Comment (zeertzjq) [reply src/nvim/window.c:5243]: I think it should be moved above the `pos_null < kv_size(buf->b_wininfo)` check.

**Pull request merged**